index.js

```diff
@@ -0,0 +1,73 @@
+'use strict'
+
+const fp = require('fastify-plugin')
+
+const { ThrottleStream } = require('./lib/throttle-stream')
+const { Readable, Stream, pipeline } = require('stream')
+
+function fastifyThrottle (fastify, options, done) {
+  options = Object.assign({}, options)
+
+  options.streamPayloads = options.streamPayloads ?? true
+  options.bufferPayloads = options.bufferPayloads || false
+  options.stringPayloads = options.stringPayloads || false
+
+  fastify.addHook('onRoute', (routeOptions) => {
+    const opts = Object.assign({}, options, routeOptions.config?.throttle)
+    if (opts.bytesPerSecond) {
+      addRouteThrottleHook(fastify, routeOptions, opts)
+    }
+  })
+  done()
+}
+
+async function addRouteThrottleHook (fastify, routeOptions, throttleOptions) {
+  const hook = 'onSend'
+  const hookHandler = throttleOnSendHandler(fastify, throttleOptions)
+  if (Array.isArray(routeOptions[hook])) {
+    routeOptions[hook].push(hookHandler)
+  } else if (typeof routeOptions[hook] === 'function') {
+    routeOptions[hook] = [routeOptions[hook], hookHandler]
+  } else {
+    routeOptions[hook] = [hookHandler]
+  }
+}
+
+function throttleOnSendHandler (fastify, throttleOpts) {
+  const bytesPerSecond = throttleOpts.bytesPerSecond
+
+  return function onSendHandler (request, reply, payload, done) {
+    if (throttleOpts.streamPayloads && payload instanceof Stream) {
+      done(null, pipeline(
+        payload,
+        new ThrottleStream({ bytesPerSecond }),
+        err => { fastify.log.error(err) }
+      ))
+      return
+    }
+    if (throttleOpts.bufferPayloads && Buffer.isBuffer(payload)) {
+      done(null, pipeline(
+        Readable.from(payload),
+        new ThrottleStream({ bytesPerSecond }),
+        err => { fastify.log.error(err) }
+      ))
+      return
+    }
+    if (throttleOpts.stringPayloads && typeof payload === 'string') {
+      done(null, pipeline(
+        Readable.from(Buffer.from(payload)),
```

Comment:
TextEncoder.encode might be faster

Reply:
would need a benchmark

Reply:
A slightly modified version of this could work: https://github.com/anonrig/node-benchmarks/blob/main/text-encoder/index.mjs

Reply:
I wonder how we can test something like this in an environment where we slow requests down.